adapters/consumable/consumable.go

```diff
@@ -0,0 +1,284 @@
+		if err == nil {
+			origin := url.URL{
+				Scheme: pageUrl.Scheme,
+				Opaque: pageUrl.Opaque,
+				Host:   pageUrl.Host,
+			}
+
+			headers.Set("Origin", origin.String())
+		}
+	}
+
+	body := bidRequest{
+		Placements:         make([]placement, len(request.Imp)),
+		Time:               a.clock.Now().Unix(),
+		IncludePricingData: true,
+		EnableBotFiltering: true,
+		Parallel:           true,
+	}
+
+	if request.Site != nil {
+		body.Referrer = request.Site.Ref // Effectively the previous page to the page where the ad will be shown
+		body.Url = request.Site.Page     // where the impression will be made
+	}
+
+	for i, impression := range request.Imp {
+
+		_, consumableExt, err := extractExtensions(impression)
+
+		if err != nil {
+			return nil, err
+		}
+
+		// These get set on the first one in observed working requests
+		if i == 0 {
+			body.NetworkId = consumableExt.NetworkId
+			body.SiteId = consumableExt.SiteId
+			body.UnitId = consumableExt.UnitId
+			body.UnitName = consumableExt.UnitName
+		}
+
+		body.Placements[i] = placement{
+			DivName:   impression.ID,
+			NetworkId: consumableExt.NetworkId,
+			SiteId:    consumableExt.SiteId,
+			UnitId:    consumableExt.UnitId,
+			UnitName:  consumableExt.UnitName,
+			AdTypes:   getSizeCodes(impression.Banner.Format), // was adTypes: bid.adTypes || getSize(bid.sizes) in prebid.js
+		}
+	}
+
+	bodyBytes, err := json.Marshal(body)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requests := []*adapters.RequestData{
+		{
+			Method:  "POST",
+			Uri:     "https://e.serverbid.com/api/v2",
+			Body:    bodyBytes,
+			Headers: headers,
+		},
+	}
+
+	return requests, nil
+}
+
+/*
+internal original request in OpenRTB, external = result of us having converted it (what comes out of MakeRequests)
+*/
+func (a *ConsumableAdapter) MakeBids(
+	internalRequest *openrtb.BidRequest,
+	externalRequest *adapters.RequestData,
+	response *adapters.ResponseData,
+) (*adapters.BidderResponse, []error) {
+
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	var serverResponse bidResponse // response from Consumable
+	if err := json.Unmarshal(response.Body, &serverResponse); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("error while decoding response, err: %s", err),
+		}}
+	}
+
+	bidderResponse := adapters.NewBidderResponse()
+	var errors []error
+
+	for impID, decision := range serverResponse.Decisions {
+		imp := getImp(impID, internalRequest.Imp)
```

Comment:
The value of imp is used only when both decision.Pricing and decision.Pricing.ClearPrice are non-nil.  Please move this code (current 206-211) into the code that runs after the check of Pricing and ClearPrice (current 215) so that it is executed only when necessary.  This will save the cost of iterating through the imps in getImp() in some cases.

Reply:
Good spot.  Change pushed